### ISSUE TITLE
Potential Non-API Based Event Sources - Rich Snippets, iCal / hCal, RSS

### ISSUE THREAD

**Issue by allella** [codeforgreenville/upstate_tech_cal_service#7]
Cross-linking back to the open data API issue history where the conversations started.

- [Rich snippets - Microdata, RDFa, JSON](https://github.com/codeforgreenville/OpenData/issues/13)
- [RSS](https://github.com/codeforgreenville/OpenData/issues/16)
- [iCal / hCal](https://github.com/codeforgreenville/OpenData/issues/14)

**Comment by Nunie123** [comment]
iCal/ics files look pretty straight forward to parse.  The problem I found was a lack of an endpoint with an ics file we can use.  @allella do you have an example of an endpoint for iCal data that this application can subscribe to?

**Comment by allella** [comment]
Yes, there are now iCal feeds and may never be any so we can close this and re-open if we get somebody who only has an iCal.

https://data.openupstate.org/organizations?city=greenville&org_status=active&event_service=ical

**Comment by allella** [comment]
Re-opening to explore the rich-snippets angle.

We have a few conferences (Grok, Restfest, Carolina Code Conf) which might not use Meetup or Eventbrite so those events don't show in the API.

We could potentially have then define their event on their websites with a rich-snippet format and pull events in that way.

**Comment by allella** [comment]
@BigBlueHat last [I checked in 2014](https://github.com/codeforgreenville/OpenData/issues/13) the supported formats were RDFa, Microdata and JSON-LD.

What's the best-practices these days?

**Comment by robertgonzales** [comment]
@allella I have access to the grok website. I can add a `.ics` file pretty easily. Is that all you need?

**Comment by allella** [comment]
Thanks @robertgonzales It's up for debate, but I'm imagining the rich snippet would be the more universal method and possibly the easiest for the average site.  Waiting on @BigBlueHat to share his wisdom on if https://rdfa.info or some other format(s) are the current suggested format(s).

We did talk about iCal and hCal support 5 years ago when this started but I'm not sure any sites have advertised an iCal other than Meetup.com and Eventbrite, for which we use the API.

**Comment by BigBlueHat** [comment]
> What's the best-practices these days?

For inline content, I'd recommend RDFa as it has the richest and most accurate tooling for expressing data within the markup (vs. microdata which results in much messier data and consequently more cleanup). Both RDFa and microdata have been supported by Google, Bing, Yahoo!, etc for a Good Long Time. 😃 RDFa is also used for systems like [Open Graph Protocol](http://ogp.me) which many CMS's already provide out the box--and which Facebook, Twitter, LinkedIn, and others use for their "card" views for links.

Google's current structured data recommendation is to [use JSON-LD](https://developers.google.com/search/docs/guides/intro-structured-data#structured-data-format). [JSON-LD](http://json-ld.org/) can be embedded in HTML using a [data block](https://www.w3.org/TR/html5/semantics-scripting.html#data-block):
```html
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "Event",
  "name": "REST Fest US East",
  "url": "http://2019.restfest.org/east/"
}
</script>
```

Depending on your setup, injecting that into your page (via your CMS, static site template, etc) can be simpler than adding attributes directly to your HTML. However, it does mean you'll likely be managing duplicated content as this content is also likely in your page already.

The [REST Fest](http://restfest.org/) uses [RDFa](http://rdfa.info/) to avoid the duplication. The dates you see on the page hold the same content the search engine will extract--so I only have to edit it in one place...which is rather nice. 😄 

Hope that helps!

**Comment by robertgonzales** [comment]
I went ahead and added JSON-LD to https://greenvillegrok.com/

**Comment by BigBlueHat** [comment]
@robertgonzales looks great and Google's [Structured Data Testing Tool](https://search.google.com/structured-data/testing-tool) likes it to. 😄 

FWIW, you can (and should) put data blocks inside `<head>`. They're inert (so they won't block DOM processing like JS does/can), and putting them earlier means search engine bots will be happier sooner--some even stop early and/or only extract content from the `<head>`.

Other than that, though, this is great example. Nicely done. 👍

**Comment by robertgonzales** [comment]
Makes sense. Thanks for the pointers.

**Comment by allella** [comment]
@oliviasculley

Here's an example with the UCLUG group.

Looks like there is JSON+LD on their event pages
![image](https://user-images.githubusercontent.com/1777776/236913359-45804969-8c44-4399-98a4-26a39e138692.png)

So, we'd mostly need to figure out how to know when there's a new event, which I suspect either needs to be from scraping the main page https://gettogether.community/upstate-carolina-linux-users-group-uclug/ unless there's some other feed on Get Together.
